vespaclient-container-plugin/src/main/java/com/yahoo/document/restapi/resource/DocumentV1ApiHandler.java

```diff
@@ -0,0 +1,617 @@
+    interface Handler {
+        ContentChannel handle(HttpRequest request, DocumentPath path, ResponseHandler handler);
+    }
+
+
+    /** Readable content channel which forwards data to a reader when closed. */
+    static class ForwardingContentChannel implements ContentChannel {
+
+        private final ReadableContentChannel delegate = new ReadableContentChannel();
+        private final Consumer<InputStream> reader;
+
+        public ForwardingContentChannel(Consumer<InputStream> reader) {
+            this.reader = reader;
+        }
+
+        @Override
+        public void write(ByteBuffer buf, CompletionHandler handler) {
+            delegate.write(buf, handler);
+        }
+
+        @Override
+        public void close(CompletionHandler handler) {
+            delegate.close(handler);
+            try (UnsafeContentInputStream in = new UnsafeContentInputStream(delegate)) {
+                reader.accept(in);
+            }
+        }
+
+    }
+
+
+    private static class DocumentOperationParser {
+
+        private static final JsonFactory jsonFactory = new JsonFactory();
+
+        private final DocumentTypeManager manager;
+
+        DocumentOperationParser(DocumentmanagerConfig config) {
+            this.manager = new DocumentTypeManager(config);
+        }
+
+        DocumentPut parsePut(InputStream inputStream, String docId) {
+            return (DocumentPut) parse(inputStream, docId, DocumentParser.SupportedOperation.PUT);
+        }
+
+        DocumentUpdate parseUpdate(InputStream inputStream, String docId)  {
+            return (DocumentUpdate) parse(inputStream, docId, DocumentParser.SupportedOperation.UPDATE);
+        }
+
+        private DocumentOperation parse(InputStream inputStream, String docId, DocumentParser.SupportedOperation operation)  {
+            return new JsonReader(manager, inputStream, jsonFactory).readSingleDocument(operation, docId);
+        }
+
+    }
+
+    private class MeasuringResponseHandler implements ResponseHandler {
+
+        private final ResponseHandler delegate;
+        private final DocumentOperationType type;
+        private final Instant start;
+
+        private MeasuringResponseHandler(ResponseHandler delegate, DocumentOperationType type, Instant start) {
+            this.delegate = delegate;
+            this.type = type;
+            this.start = start;
+        }
+
+        @Override
+        public ContentChannel handleResponse(Response response) {
+            switch (response.getStatus() / 100) {
+                case 2: metrics.reportSuccessful(type, start); break;
+                case 4: metrics.reportFailure(type, DocumentOperationStatus.REQUEST_ERROR); break;
+                case 5: metrics.reportFailure(type, DocumentOperationStatus.SERVER_ERROR); break;
+            }
+            return delegate.handleResponse(response);
+        }
+
+    }
+
+    private Metric.Context contextFor(Request request, Map<String, String> extraDimensions) {
```

Comment:
This logic is duplicated from one of the threaded request handler base classes.

Reply:
Yes. I suppose I could move it somewhere else, where both of them have access to it.